From 5626fdcfdaf562c7782a2c79f10a57b46b6087c4 Mon Sep 17 00:00:00 2001
From: Paul Arthur <paul.arthur@flowerysong.com>
Date: Thu, 17 May 2018 01:41:53 -0400
Subject: [PATCH] Fix debuginfo downloading

---
 src/client-python/reportclient/yumdebuginfo.py | 2 --
 1 file changed, 2 deletions(-)

diff --git a/src/client-python/reportclient/yumdebuginfo.py b/src/client-python/reportclient/yumdebuginfo.py
index 0ddabd92..bdbad17e 100644
--- a/src/client-python/reportclient/yumdebuginfo.py
+++ b/src/client-python/reportclient/yumdebuginfo.py
@@ -77,8 +77,6 @@ class YumDebugInfoDownload(DebugInfoDownload):
         super(YumDebugInfoDownload, self).__init__(cache, tmp, repo_pattern, keep_rpms, noninteractive)
 
         self.base = YumBase()
-        if not releasever is None:
-            self.base.conf.substitutions['releasever'] = releasever
 
     def initialize_progress(self, updater):
         self.progress = YumDownloadCallback(updater)
-- 
2.14.3

