diff -ur redis-6.0.5/src/Makefile redis-6.0.5-/src/Makefile
--- redis-6.0.5/src/Makefile	2020-06-09 06:19:30.000000000 -0400
+++ redis-6.0.5-/src/Makefile	2020-06-16 16:24:33.219036382 -0400
@@ -186,9 +186,8 @@
 endif
 
 ifeq ($(MALLOC),jemalloc)
-	DEPENDENCY_TARGETS+= jemalloc
-	FINAL_CFLAGS+= -DUSE_JEMALLOC -I../deps/jemalloc/include
-	FINAL_LIBS := ../deps/jemalloc/lib/libjemalloc.a $(FINAL_LIBS)
+	FINAL_CFLAGS+= -DUSE_JEMALLOC -DJEMALLOC_NO_DEMANGLE -I/usr/include/jemalloc
+	FINAL_LIBS+= -ljemalloc
 endif
 
 ifeq ($(BUILD_TLS),yes)
